src/librustc_codegen_ssa/mir/constant.rs

```diff
@@ -40,31 +40,36 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         &mut self,
         constant: &mir::Constant<'tcx>,
     ) -> Result<ConstValue<'tcx>, ErrorHandled> {
-        match constant.literal.val {
+        let const_ = match constant.literal.val {
             ty::ConstKind::Unevaluated(def_id, substs, promoted) => {
                 let substs = self.monomorphize(&substs);
-                self.cx
-                    .tcx()
-                    .const_eval_resolve(ty::ParamEnv::reveal_all(), def_id, substs, promoted, None)
-                    .map_err(|err| {
-                        if promoted.is_none() {
-                            self.cx
-                                .tcx()
-                                .sess
-                                .span_err(constant.span, "erroneous constant encountered");
-                        }
-                        err
-                    })
+                ty::ConstKind::Unevaluated(def_id, substs, promoted)
             }
+            ty::ConstKind::Value(value) => ty::ConstKind::Value(value),
+            ty::ConstKind::Param(_) => self.monomorphize(&constant.literal).val,
+            _ => span_bug!(constant.span, "encountered bad Const in codegen: {:?}", constant),
+        };
```

Comment:
My suggestion was to have `let const_ = self.monomorphize(&constant.literal);` instead of this `match`, btw, no need to check cases.